happybase/client.py

```diff
@@ -28,41 +43,70 @@ def _req(self, _api, *args, **kwargs):
             except (TApplicationException, socket_error, TTransportException) as exc:
                 logger.exception("Got exception")
                 while True:
                     interval = retries.popleft() if retries else interval
                     logger.info("Sleeping for %d seconds", interval)
-                    sleep(interval)
-                    logger.info("Trying to reconnect")
-                    try:
-                        self._connection._refresh_thrift_client()
-                        self._connection.open()
-                        client = super(RecoveringClient, self._connection.client)
-                        logger.debug("New client is initialized")
-                    except TTransportException:
-                        logger.exception("Got exception, while trying to reconnect. Continuing")
-                        pass
-                    else:
+                    time.sleep(interval)
+                    client = self._reopen_connection()
+                    if client:
                         break
                 if no_retry:
                     raise exc
             finally:
-                now = time()
-                elapsed = (now - t) * 1E+3
+                now = utc_timestamp()
+                elapsed = int((now - t) * 1E+3)
                 logger.debug("RPC call to %s took %d ms", _api, elapsed)
-                self._resp_times.append(elapsed)
+                # XXX we're going to send all timings since last get_stats() call,
+                # so each deque stores pairs of (utc timestamp, timing in ms)
+                self._response_timings[_api].append((now, elapsed))
                 if self._next_log_time < now:
                     self._log_response_times()
                     self._next_log_time = now + self._log_interval
 
+    def _reopen_connection(self):
+        """Reopen connection to HBase.
+
+        Method should return a client instance or None on failed attempt.
+        """
+        logger.info("Trying to reconnect")
+        try:
+            self._connection._refresh_thrift_client()
+            self._connection.open()
+            client = super(RecoveringClient, self._connection.client)
+            logger.debug("New client is initialized")
+        except TTransportException:
+            logger.exception("Got exception, while trying to reconnect. Continuing")
+            pass
+        else:
+            return client
+
     def _log_response_times(self):
         logger.debug("client id %d response times (ms) statistics", id(self))
-        logger.debug("mean: %f stdev: %f median: %f count: %d",
-                     np.mean(self._resp_times),
-                     np.std(self._resp_times),
-                     np.median(self._resp_times),
-                     len(self._resp_times))
\ No newline at end of file
+        for api_entry in self._response_timings:
+            logger.debug("%s mean: %f stdev: %f median: %f count: %d",
```

Comment:
Having some doubts about this, it's more relevant to print stats per function type, not summary of all values, what do you think? At the same point it means more logs, wonder if it's appropriate.